hydrogym/ts.py

```diff
@@ -49,7 +50,14 @@ def enlist_controls(self, control):
 
 
 class IPCS(TransientSolver):
-    def __init__(self, flow: FlowConfig, dt: float, debug=False, **kwargs):
+    def __init__(
+        self,
+        flow: FlowConfig,
+        dt: float,
+        control_method="direct",
```

Comment:
I liked your original idea of "velocity" or "force" being the options - I think "direct" and "indirect" might be less obvious.

Reply:
My only concern is will this be abstract able to other Flow environments?

Reply:
Actually I guess we could just throw all of "update_controls" into the flow config. Then, put this code block:

`
def update_controls(self, control):
            """Add a damping factor to the controller response

            If actual control is u and input is v, effectively
                du/dt = (1/tau)*(v - u)
            """
            for (u, v) in zip(self.control, control):
                u = u + (self.dt / self.flow.TAU) * (v - u)
            # for direct control of the state, the input is the desired state
            self.state = self.control
`

into the core.py so that all environments have access to the old way of updating state.

Then this could then be overridden in each individual environment so that we could use a keyword more environment specific. Then we would have update_environment in the cylinder env look like:

`
def update_controls(self, control):
        if self.control_method == "indirect":
            # Update state
            self.flow.update_state(control, self.dt)

        else:
            """Add a damping factor to the controller response

            If actual control is u and input is v, effectively
                du/dt = (1/tau)*(v - u)
            """
            for (u, v) in zip(self.control, control):
                u = u + (self.dt / self.flow.TAU) * (v - u)
            # for direct control of the state, the input is the desired state
            self.state = self.control
`

except with the appropriate variables called where necessary.


I will finish all other changes before coming back to this one since this seems like the most aggressive change.

Reply:
PS idk why my code is formatting like that, but I hope the just of what I'm saying is clear without the examples

Reply:
The other option for easier Python formatting is also to just use [YAPF](https://github.com/google/yapf) which will do its best to format your code in "good-faith" :)

Reply:
So the more I thought about this, the more I believe that removing "update_controls/update_state" from TS and putting it in the flow config is necessary since (it is my understanding that) each flow environment will need to be controlled via it's own set of differential equations. Thus, "update_controls/update_state" is inherently unique to each flow environment.

I do understand that this is a pretty drastic change to the structure of the project, so if it interferes with development at the moment, we could push this to a later date and I can open a git issue with low priority. If this seems worth it to add now though, I was able to finish fully decoupling TS and Flow Config in my most recent push to this branch. All tests work for the cylinder and pinball. I'm just running into some issues with the "integrate" test in the cavity test suite which seems like a separate issue (via #38 ). Let me know what you think @jcallaham since I know you are still developing the backwards step and I don't want to throw a wrench in your work if it's too much for right now.

Reply:
Awesome! This should be resolved then in commit "a4a6bf4" for the cylinder, pinball and cavity!